### DISCUSSION TITLE
How to implement refresh token in remix?

### DISCUSSION THREAD

**Discussion by ShourovRoy** [remix-run/remix#6252] [Q&A]
I am using separate  express server and it has refresh token function. In remix I am able to refresh the tokens using axios interceptors but from interceptor failed to create new createUserSession with new access and refresh token.

I am hiding the access and refresh token inside the remix sessions and extracting them inside axios interceptors but from there I stuck because the createUserSession not working there anymore.

Please help me

**Comment by devraj** [comment]
I have been trying to fin da solution to this as well. I found [this blog post](https://sergiodxa.com/articles/working-with-refresh-tokens-in-remix) that outlines a way to do this.

I am yet to find a way to register an `interceptor` properly, there are no lifecycle methods (that I can find at the moment) that would allow you to do this conveniently. Part of the issue is that the only logical place to store the `token` is in a remix session which is only available as part of a request. 

I had an initial idea of applying the token in the `root` loader, this has a variant rate of success as the loader does not get called (I guess this is expected behaviour).

@sergiodxa's blog post I linked to mentions that ideally a `pre` fetch function would be ideal.

One of my attempts tried to add the interceptor in the loader, but that will make the interceptor register multiple times as `remix` managed the lifecycle of the `loader` calls.

Failing everything I guess we have to do wha the blog post suggests and add the headers on each loader or action where we need it.

I am aiming to spend some more time on this shortly. You can monitor [my progress here](https://github.com/anomaly/psychedelic-stack). I will post on this thread if I am successful.

**Reply by devraj** [reply]
See also https://github.com/remix-run/remix/discussions/6713

**Reply by sergiodxa** [reply]
If you're ok with doing code in your HTTP server, you could do the token refresh there, then pass the access token to the app with getLoadContext.

**Reply by nativeTiger** [reply]
@sergiodxa It will be helpful if you could provide the working example for doing code in HTTP server.
